src/App.tsx

```diff
@@ -1,26 +1,266 @@
+                  * click delete markings
+                  */
+            }
+        }
+        this.setState({
+            values: newValues,
+        });
+    }
+
+/*     convertUndefinedValues(): (SquareValue | null)[] {
+        let list = [];
+        for (var i = 0; i < 81; i++) {
+            if (this.props.values[i] == undefined) {
+                list.push(null);
+            } else {
+                list.push(this.props.values[i] as SquareValue);
+            }
+        }
+        return list;
+    } */
+
+    public render() {
+        this.state.permanentValues = this.getPermanentValues();
+        return(
+            <div className="game">
+                <div className="board">{this.renderSquares()}</div>
+                <div className="button-box">
+                    <div className="button-box-top">
+                        <div className="button-col">
+                            <NormalButton/>
+                            <CornerButton/>
+                            <UndoButton/>
+                            <RedoButton/>
+                        </div>
+                        <Numpad
+                        onClickNum={(i: SquareValue) => {this.normalMarkSelectedSquares(i); }}
+                        onClickDel={() => {this.deleteSelectedSquares(); }}
+                        />
+                    </div>
+                    <div className="button-box-bot">
+                        <Button variant="contained" color="primary">
+                          R E S T A R T
+                        </Button>
+                        <Button variant="contained" color="secondary">
+                          C H E C K
+                        </Button>
+                    </div>
+                </div>
+            </div>
+        );
+    }
+
+    public renderSquares() {
+        const list = [];
+        for (let i = 0; i < 81; i++) {
+            list.push(this.renderSquare(i));
+        }
+        return list;
+    }
+
+    public renderSquare(i: SquareAddress) {
+        const isHighlighted = this.state.highlights.includes(i);
+        const isContradicting = this.state.contradicts.includes(i);
+        const marks = i % 9 + 1 == 7 ? [i % 1 as SquareValue, i % 5 + 1 as SquareValue, i % 3 + 1 as SquareValue, i % 8 + 1 as SquareValue] : [];
+        return <Square value={this.state.values[i]}
+        isPermanent={this.state.permanentValues.includes(i)} isHighlighted = {isHighlighted} isContradicting={isContradicting}
+        markings={marks} onClick={() => { this.toggleSelectSquare(i); }}/>;
+    }
+}
+
+interface SquareProps {
+    isPermanent: boolean;
+    isHighlighted: boolean;
+    isContradicting: boolean;
+    value: SquareValue | null;
+    markings: SquareValue[];
+    onClick: () => void;
+}
+
+type SquareValue = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 ;
+
+class Square extends React.Component<SquareProps, {}> {
+    public getMarks(): boolean[] {
```

Comment:
Please rewrite this as:

```js
public isMarked(squareValue: SquareValue): boolean
```

Reply:
On second thought, maybe replace with a `renderMarking(squareValue: SquareValue)` method.